From 209eb19e399a2c9f8d9ca609fc5a772a149d9bd8 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:19:28 -0400
Subject: [net] netfilter: bridge: add reject support

Message-id: <20141021161928.22888.16654.stgit@localhost.localdomain>
Patchwork-id: 97979
O-Subject: [RHEL7 net PATCH 082/124] netfilter: bridge: add reject support
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 85f5b3086a04c459f9147859fcbf7bdc7578c378

 commit 85f5b3086a04c459f9147859fcbf7bdc7578c378
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Fri Jun 27 13:36:11 2014 +0200

    netfilter: bridge: add reject support

    So you can reject IPv4 and IPv6 packets from bridge tables. If the ether
    proto is now known, default on dropping the packet instead.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/bridge/netfilter/Kconfig b/net/bridge/netfilter/Kconfig
index 3baf29d..45b3223 100644
--- a/net/bridge/netfilter/Kconfig
+++ b/net/bridge/netfilter/Kconfig
@@ -15,6 +15,12 @@ config NFT_BRIDGE_META
  help
    Add support for bridge dedicated meta key.
 
+config NFT_BRIDGE_REJECT
+ tristate "Netfilter nf_tables bridge reject support"
+ depends on NFT_REJECT && NFT_REJECT_IPV4 && NFT_REJECT_IPV6
+ help
+   Add support to reject packets.
+
 endif # NF_TABLES_BRIDGE
 
 menuconfig BRIDGE_NF_EBTABLES
diff --git a/net/bridge/netfilter/Makefile b/net/bridge/netfilter/Makefile
index 6f2f394..ee3c586 100644
--- a/net/bridge/netfilter/Makefile
+++ b/net/bridge/netfilter/Makefile
@@ -4,6 +4,7 @@
 
 obj-$(CONFIG_NF_TABLES_BRIDGE) += nf_tables_bridge.o
 obj-$(CONFIG_NFT_BRIDGE_META)  += nft_meta_bridge.o
+obj-$(CONFIG_NFT_BRIDGE_REJECT)  += nft_reject_bridge.o
 
 obj-$(CONFIG_BRIDGE_NF_EBTABLES) += ebtables.o
 
diff --git a/net/bridge/netfilter/nft_reject_bridge.c b/net/bridge/netfilter/nft_reject_bridge.c
new file mode 100644
index 0000000..ee3ffe9
--- /dev/null
+++ b/net/bridge/netfilter/nft_reject_bridge.c
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2014 Pablo Neira Ayuso <pablo@netfilter.org>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/netlink.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter/nf_tables.h>
+#include <net/netfilter/nf_tables.h>
+#include <net/netfilter/nft_reject.h>
+
+static void nft_reject_bridge_eval(const struct nft_expr *expr,
+     struct nft_data data[NFT_REG_MAX + 1],
+     const struct nft_pktinfo *pkt)
+{
+ switch (eth_hdr(pkt->skb)->h_proto) {
+ case htons(ETH_P_IP):
+  return nft_reject_ipv4_eval(expr, data, pkt);
+ case htons(ETH_P_IPV6):
+  return nft_reject_ipv6_eval(expr, data, pkt);
+ default:
+  /* No explicit way to reject this protocol, drop it. */
+  data[NFT_REG_VERDICT].verdict = NF_DROP;
+  break;
+ }
+}
+
+static struct nft_expr_type nft_reject_bridge_type;
+static const struct nft_expr_ops nft_reject_bridge_ops = {
+ .type  = &nft_reject_bridge_type,
+ .size  = NFT_EXPR_SIZE(sizeof(struct nft_reject)),
+ .eval  = nft_reject_bridge_eval,
+ .init  = nft_reject_init,
+ .dump  = nft_reject_dump,
+};
+
+static struct nft_expr_type nft_reject_bridge_type __read_mostly = {
+ .family  = NFPROTO_BRIDGE,
+ .name  = "reject",
+ .ops  = &nft_reject_bridge_ops,
+ .policy  = nft_reject_policy,
+ .maxattr = NFTA_REJECT_MAX,
+ .owner  = THIS_MODULE,
+};
+
+static int __init nft_reject_bridge_module_init(void)
+{
+ return nft_register_expr(&nft_reject_bridge_type);
+}
+
+static void __exit nft_reject_bridge_module_exit(void)
+{
+ nft_unregister_expr(&nft_reject_bridge_type);
+}
+
+module_init(nft_reject_bridge_module_init);
+module_exit(nft_reject_bridge_module_exit);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Pablo Neira Ayuso <pablo@netfilter.org>");
+MODULE_ALIAS_NFT_AF_EXPR(AF_BRIDGE, "reject");
-- 
1.7.1